systems/trajectory_optimization/dircon_kinematic_data_set.h

```diff
@@ -59,25 +59,49 @@ class DirconKinematicDataSet {
   // elements
   class CacheHasher {
    public:
-    std::size_t operator()(const CacheKey& key) const {
-      std::size_t hash = 0;
+    std::size_t operator()(const CacheKey<double>& key) const {
+      std::size_t ret = 0;
       for (int i = 0; i < key.state.rows(); i++) {
-        hash = (hash << 1) ^ std::hash<double>{}(key.state(i));
+        ret = (ret << 1) ^ std::hash<double>{}(key.state(i));
       }
       for (int i = 0; i < key.input.rows(); i++) {
-        hash = (hash << 1) ^ std::hash<double>{}(key.input(i));
+        ret = (ret << 1) ^ std::hash<double>{}(key.input(i));
       }
       for (int i = 0; i < key.forces.rows(); i++) {
-        hash = (hash << 1) ^ std::hash<double>{}(key.forces(i));
+        ret = (ret << 1) ^ std::hash<double>{}(key.forces(i));
       }
-      return hash;
+      return ret;
+    }
+
+    std::size_t operator()(const CacheKey<drake::AutoDiffXd>& key) const {
+      using std::hash;
+      std::size_t ret = 0;
+      for (int i = 0; i < key.state.rows(); i++) {
+        ret = (ret << 1) ^ hash<double>{}(key.state(i).value());
+        for (int j = 0; j < key.state(i).derivatives().size(); j++) {
+          ret = (ret << 1) ^ hash<double>{}(key.state(i).derivatives()[j]);
+        }
```

Comment:
Just one comment -- I think this might not be necessary, since we probably don't get states with the same values but different derivatives? (same comment for input and force)

Reply:
I take the above comment back. Apparently it's necessary to encode the derivatives into the key, since SNOPT failed to find a solution after I removed it.